scaffoldmaker/meshtypes/meshtype_3d_ellipticcylinder1.py

```diff
@@ -0,0 +1,476 @@
+        if options['Height'] < 0.0:
+            options['Height'] = 0.0
+        if options['Thickness ratio'] <= 0.0:
+            options['Thickness ratio'] = 0.01
+        elif options['Thickness ratio'] > 1.0:
+            options['Thickness ratio'] = 0.99
+
+    @classmethod
+    def generateBaseMesh(cls, region, options):
+        """
+        Generates the base tricubic Hermite mesh for this MeshType. See also generateMesh().
+        :param region: Zinc region to define model in. Must be empty.
+        :param options: Dict containing options. See getDefaultOptions().
+        :return: None
+        """
+        elementsCountAround = options['Number of elements around']
+        elementsCountUp = options['Number of elements up']
+        elementsCountThroughWall = options['Number of elements through wall']
+        useCrossDerivatives = options['Use cross derivatives']
+        axisA = options['Major diameter']
+        axisB = options['Minor diameter']
+        height = options['Height']
+        useCubicHermiteThroughWall = True
+
+        a = identity(4)
+
+        zero = [0.0, 0.0, 0.0]
+
+        """ Get nodes along the centre line """
+        cx = [zero, [0.0, 0.0, height]]
+        cd1 = [[0.0, 0.0, height / elementsCountUp], [0.0, 0.0, height / elementsCountUp]]
+        cd2 = [[0.0, axisB / elementsCountThroughWall, 0.0], [0.0, axisB / elementsCountThroughWall, 0.0]]
+        cd3 = [[axisA / elementsCountThroughWall, 0.0, 0.0], [axisA / elementsCountThroughWall, 0.0, 0.0]]
+
+        t = options['Thickness ratio']
+        t2 = [1.0 - options['Thickness ratio']] * 2
+        t2d = [0.0, 0.0]
+        t3 = [1.0 - options['Thickness ratio']] * 2
+        t3d = [0.0, 0.0]
+
+        sx, sd1, se, sxi, _ = sampleCubicHermiteCurves(cx, cd1, elementsCountUp)
+        sd2 = interpolateSampleLinear(cd2, se, sxi)
+        sd3 = interpolateSampleLinear(cd3, se, sxi)
+        st2 = interpolateSampleLinear(t2, se, sxi)
+        st2d = interpolateSampleLinear(t2d, se, sxi)
+        st3 = interpolateSampleLinear(t3, se, sxi)
+        st3d = interpolateSampleLinear(t3d, se, sxi)
+
+        # Find unit normals and binormals at each sample points
+        sNormal = []
+        sBinormal = []
+
+        # Set up normal and binormal for first frame
+        prevUnitTangent = normalise(sd1[0])
+        if magnitude(crossproduct3(prevUnitTangent, [0.0, 0.0, 1.0])) > 0.0:
+            prevBinormal = crossproduct3(prevUnitTangent, [0.0, 0.0, 1.0])
+        else:
+            prevBinormal = crossproduct3(prevUnitTangent, [0.0, -1.0, 0.0])
+        prevUnitBinormal = normalise(prevBinormal)
+        prevUnitNormal = crossproduct3(prevUnitBinormal, prevUnitTangent)
+        sNormal.append(prevUnitNormal)
+        sBinormal.append(prevUnitBinormal)
+
+        # Step through central line and rotate central line axes to align tangent
+        # to tangent from previous level
+        for n in range(1, elementsCountUp + 1):
```

Comment:
I'm not sure what this is achieving and why the whole rotation matrix stuff is needed. We will need to discuss.